### PR TITLE
buildsystem: support choosing default linker

### PR DESCRIPTION
**Pull request by HiassofT**
This is an alternative to #6922 allowing more flexible linker configuration.

Default linker can be set with DEFAULT_LINKER in options.

Packages can influence linker selection both by positive and/or negative PKG_BUILD_FLAGS, eg +bfd or -gold.

Positive build flags take priority over the default linker so eg DEFAULT_LINKER="gold" and PKG_BUILD_FLAGS="+bfd" will select bfd.

Negative flags mean a specifiy linker should not be used, eg -gold prevent using gold.

If the default linker is disabled via a build flag then any other available linker will be used (typically falling back to bfd which is always present).

Optional linkers like gold have to be enabled with eg GOLD_SUPPORT="yes" in options. If an optional linker is not enabled it won't be a candidate for linker selection. So eg "+mold" will have no effect if MOLD_SUPPORT isn't set to "yes".

### CODE REVIEW

**Comment by HiassofT** [comment]
Build test just finished fine so it's ready for review.

Here is some more info about the rationale behind this PR:

While it doesn't have much effect on our current setup it will help us adding and testing mold linker support (or maybe other linkers in the future).

For adding mold support we have to keep several use cases in mind:

First of all mold needs gcc version 10 or later on the build host which isn't available on all current distributions. So we can't enable mold build by default ATM which means our build system needs to cope with mold support being optional.

This can easily be achieved by adding mold to `all_linkers` in the get_target_linker function and chainging `linker_candidates` to only include gold and bfd in addition to the default linker. eg
```
local all_linkers="mold gold bfd"
local linker_candidates="${DEFAULT_LINKER:-bfd} gold bfd"
```

This allows testing individual packages with mold by setting the `+mold` explicit opt-in build flag and also allows us to test building all packages with mold by changing the DEFAULT_LINKER option.

During these test we most likely will find out that some packages will fail to build with mold - we can then easily disable mold support for the individual package via the `-mold` opt-out flag.

For packages that fail to build both with mold and gold we have 2 options: either use `-gold -mold` flags or `+bfd`. There are pros and cons to both approaches so it'll mainly be a design decision which one to choose.

If we are 100% certain we'll never want to use anything else than bfd then we can add `+bfd`. The drawback though is that if in the future we'd like to test with yet another linker (maybe lld) we have to change the package.mk file - changing default linker etc won't help here as that specific linker is forced.

There may be other reasons for forcing a specific linker but in general I guess opt-out will be the better approach.

**Comment by HiassofT** [comment]
I've switched kodi to use the get_target_linker helper and did some more test.

Disabling GOLD_SUPPORT or switching default linker to bfd is not a good idea, kodi then fails to link samba libs - but the same happens with  current master if GOLD_SUPPORT is disabled so is unrelated to this PR

**Comment by SupervisedThinking** [comment]
Well maybe I lack some kind of imagination but I still don't see the purpose of having tri-state linker flags. From my perspective, since I & others have done it this way before, adding an additional linker is done by roughly those step:
- update the way the build system handles the linker (basically this PR)
- add & set a default linker
- add the linker package & update the build opts to mandatory/conditionally build another linker
- build the distro and fix packages which don't link with mold gold etc.
- re-check the "broken" packages with every gcc/package version bump

E.g. once we figure out how to speed up mold then mold is set as default linker. So only few packages have to be fixed by `+bfd`/`+gold`, if we still default to gold & mold is an opt-in it doesn't make sense at all to add +mold flags to packages anywhere. If there is a package which fails to link with gold/mold we now rather add `-gold -mold` instead of `+bfd` maybe even `-gold -mold -lld`? Well IMHO thats overly complicated and I don't see the problem which is fixed by this tri-state flags 🤔 Also it's quite clear what we use to link & instead of just listing what we don't use.

As long as we don't add another linker it will still default to gold as we did in the past since the vast majority links with gold anyway. And if we swap gold & mold then probably because almost all link with mold(from my tests they do). The number of packages which need a flag to overwrite the linker flags are rather few and as you can see in https://github.com/LibreELEC/LibreELEC.tv/pull/6914 it's certainly no final state, with every package bump or gcc bump we still have to / should check if the package finally builds with gold or any other linker.

So sure we can add this PR - but I clearly don't understand the purpose or problem which it fixes.

**Comment by HiassofT** [comment]
Consider the following - not too unlikely - usecase:

We are going to add mold as an optional linker to our buildsystem but we keep the default at gold for performance reason.

Only a few package might benefit from mold so you might want to flag those to use mold (if available).

Most other packages will be built with the default gold linker.

A few packages though won't be able to build eg with gold so you want to flag those to not use that linker.

This means you already have 3 states: prefer a linker over default, use default, prevent a linker from being used.

This PR already caters for this usecase, eg add "+mold" or "-gold" - if mold is not available, eg MOLD_SUPPORT is turned off then "+mold" will have no effect and some other linker will be picked.

In addition to that you will probably want to test (or permantently) build with some different linker, eg mold - for whatever reasons.

In this case you just have to change DEFAULT_LINKER in options and rebuild - packages without an explicit + linker flag will automatically use them - hence negative flagging of linkers is in general the better solution for packages as it means they actually can use the default linker - which is the main purpose of having a DEFAULT_LINKER setting :-)

**Comment by SupervisedThinking** [comment]
> Consider the following - not too unlikely - usecase:

- if gold is still the default linker & mold is build conditionally we don't need extra opt-out flags.
- if gold is still the default linker & mold is build mandatory we you would simply pass +mold or +bfd to the packages which won't link with gold
- if mold is the default linker you would pass +bfd +gold to any package which does not link with mold

This only would only make sense if we keep gold as default, conditionally build mold (so it's not always available) and add `+mold` to some packages in master which could be build faster if mold is available (but those flags are then just superflous if mold is conditionally only available).Then again this is no standard build but rather some downstream hack trial & error build.

Either there is an overall & general benefit if it comes to build times or not because it's quite unlikely that the speed up of some packages safes more time the build of mold & dependencies does.

If there is stuff which won't build with mold you could simply pass +gold even if mold is optional, then gold would be enforced in any case, same for +bfd.

So again I don't understand the actual usecase. Yes in theory if we opt in mold but mandatory add conditionally supported +mold flags to some packages then we could run into trouble but then you could still add a fallback e.g. to the bfd flag which is probably a oneliner here https://github.com/LibreELEC/LibreELEC.tv/pull/6875/files#diff-336bacf0619824ca0f2d68dd2464aa2d1deb0fb366500485591cc8b837f7fdabR293

From my perspective switching the default linker only makes sense if the distro builds faster. If this is the case we just fix the packages which are troublesome by the workaround +bfd +gold & report them to the devs. So yes add it if this fixes some theoretical problems but I still don't understand the real usecase 😄

**Comment by HiassofT** [comment]
Well, if you pinned a package to use eg +bfd or +mold then how would you test building with a different linker?

Yes, exactly, you have to start touching the package.mk file.

I don't see us switching to mold as default any time soon, not only because of the gcc < 10 issue but simply because it hasn't got the extensive testing with LE, it is quite new in general and there are still issues being ironed out.

So for quite some time we'll have to test that and need to quickly switch between eg gold and mold. Flipping the single DEFAULT_LINKER option is easy to do that, changing package.mk files is tedious (and not necessary).

Another purpose of this PR is also to provide a simple, centralized and well defined logic for linker selection - which is also needed for packages like kodi which currently have their own GOLD_SUPPORT check (which will become tedious once we add mold support).

This makes it easy to change or adapt the logic at a single place - and I think the current implementation, which is just 2 simple loops, is easy enough to understand and with the concept of default as the base, possibility to prefer something else over default and possibility to not use some linkers it's flexible enough for pretty much all usecases.

It's also very easy to extend to yet another linker, and definitely more maintainable than starting with nested if/else constructs.

**Comment by SupervisedThinking** [comment]
Okay I can imagine a use case now... 
- lets say we add mold & linker-nx
- we know a package links fine with gold & linker-nx but not with mold
- gold & linker-nx link with **exact** the same speed so we don't want to prefer one linker over the other

Then `-mold` would make sense over +gold / +linker-nx if we change the DEFAULT_LINKER to either gold or linker-nx & don't have to change the package then. Well seems quite unlikely to me that this will ever happen since we already now bfd is slower than gold and no real competitor.

**Comment by HiassofT** [comment]
Let me rephrase it in a different way:

This PR preserves the way linker compatibility is configured in packages - via opt-out flags.

In addition to that it adds two more features:

- the possibility to change the default linker
- the possibility to prefer a different linker than the default (opt-in flags)

It doesn't mean you have to use any of the additional features - though they will come very handy.

Why would you want to change the logic and switch a bunch of package.mk files from -gold to eg +bfd? Sure, you can do that to hard-pin them (which might be the cause for glibc) but otherwise you loose the possibility to easily test eg with mold.

So I think it's good to have both variants.

The get_target_linker helper takes care about all that - and as it also checks package flags, presence of linker etc it makes sense to use that in packages, too, instead of  DEFAULT_LINKER (or GOLD_SUPPORT etc).

Kodi's package.mk then only needs to take care about setting the correct kodi cmake options (which will typically override or C/LD_FLAGS).

**Review by SupervisedThinking — APPROVED** [review]
